lib/ParseSIL/ParseSIL.cpp

```diff
@@ -885,6 +887,75 @@ void SILParser::convertRequirements(SILFunction *F,
   }
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+/// Parse a SIL constant value (one of the following categories):
+/// - An integer literal (42).
+/// - A floating point literal (3.14).
+/// - A metatype (the instance type is parsed) ($Float).
+/// - An aggregate ([1, 2, 3]).
+/// Returns true on error.
+static bool parseSymbolicValue(SymbolicValue &value, SILParser &SP,
+                               SILBuilder &B) {
+  auto &P = SP.P;
+  auto &allocator = B.getModule().getASTContext().getAllocator();
+  // Handle integer literals.
+  if (P.Tok.is(tok::integer_literal)) {
+    APInt intValue;
+    if (SP.parseInteger(intValue, diag::sil_expected_graph_op_attr_value))
+      return true;
+    value = SymbolicValue::getInteger(intValue, allocator);
+    return false;
+  }
+  // Handle floating point literals.
+  if (P.Tok.is(tok::floating_literal)) {
+    double floatValue;
+    P.Tok.getText().getAsDouble(floatValue);
+    P.consumeToken(tok::floating_literal);
+    value = SymbolicValue::getFloat(APFloat(floatValue), allocator);
+    return false;
+  }
+  // Handle string literalss.
+  if (P.Tok.is(tok::string_literal)) {
+    // TODO: Uncomment when `getStringValue` is implemented.
+    // StringRef rawString = P.Tok.getText().drop_front().drop_back();
+    // value = SymbolicValue::getStringValue(rawString, allocator);
+    // return false;
+    P.diagnose(P.Tok, diag::sil_unhandled_graph_op_attr_value);
+    return true;
+  }
+  // Handle metatypes (the instance type is parsed).
+  if (P.Tok.is(tok::sil_dollar)) {
+    SILType temp;
+    if (SP.parseSILType(temp))
+      return true;
+    auto metatype = CanMetatypeType::get(temp.getSwiftRValueType());
+    value = SymbolicValue::getMetatype(metatype);
+    return false;
+  }
+  // Handle aggregate literals.
+  if (P.consumeIf(tok::l_square)) {
+    SmallVector<SymbolicValue, 8> elements;
+    while (true) {
```

Comment:
This doesn't look quite right.  It was accept  [123,].  Comma separated lists are annoying, and it is doubly annoying that the SIL parser doesn't have a helper for this.  Please search for tok::comma to see how other loops in the SIL parser are implemented and follow one of their patterns, thanks!

Reply:
Actually, it looks like Parser::parseList might be able to do this.  SILParser has a "Parser" member named P.

Reply:
Done. I used `parseList` here and in the other places.